packages/dev/core/src/Meshes/greasedLineMesh.ts

```diff
@@ -0,0 +1,535 @@
+                        object: this,
+                    });
+                    if (firstOnly) {
+                        return intersects;
+                    }
+                }
+            }
+            i = l;
+        }
+
+        return intersects;
+    }
+
+    private _initGreasedLine() {
+        this._vertexPositions = [];
+        this._counters = [];
+        this._previous = [];
+        this._next = [];
+        this._side = [];
+        this._indices = [];
+        this._uvs = [];
+    }
+
+    private _updateRaycastBoundingInfo() {
+        const boundingInfo = this.getBoundingInfo();
+        this._boundingSphere.reConstruct(boundingInfo.minimum, boundingInfo.maximum, this._matrixWorld);
+    }
+
+    private static _CompareV3(positionIdx1: number, positionIdx2: number, positions: number[]) {
+        const arrayIdx1 = positionIdx1 * 6;
+        const arrayIdx2 = positionIdx2 * 6;
+        return positions[arrayIdx1] === positions[arrayIdx2] && positions[arrayIdx1 + 1] === positions[arrayIdx2 + 1] && positions[arrayIdx1 + 2] === positions[arrayIdx2 + 2];
+    }
+
+    private static _CopyV3(positionIdx: number, positions: number[]) {
+        const arrayIdx = positionIdx * 6;
+        return [positions[arrayIdx], positions[arrayIdx + 1], positions[arrayIdx + 2]];
+    }
+
+    private _preprocess(positions: number[]) {
+        const l = positions.length / 6;
+
+        let v: number[] = [];
+
+        const previous = [];
+        const next = [];
+        const side = [];
+        const uvs = [];
+
+        if (GreasedLineMesh._CompareV3(0, l - 1, positions)) {
+            v = GreasedLineMesh._CopyV3(l - 2, positions);
+        } else {
+            v = GreasedLineMesh._CopyV3(0, positions);
+        }
+        previous.push(v[0], v[1], v[2]);
+        previous.push(v[0], v[1], v[2]);
+
+        for (let j = 0; j < l; j++) {
+            side.push(1);
+            side.push(-1);
+
+            // uvs
+            uvs.push(j / (l - 1), 0);
+            uvs.push(j / (l - 1), 1);
+
+            if (j < l - 1) {
+                v = GreasedLineMesh._CopyV3(j, positions);
+                previous.push(v[0], v[1], v[2]);
+                previous.push(v[0], v[1], v[2]);
+            }
+            if (j > 0) {
+                v = GreasedLineMesh._CopyV3(j, positions);
+                next.push(v[0], v[1], v[2]);
+                next.push(v[0], v[1], v[2]);
+            }
+        }
+
+        if (GreasedLineMesh._CompareV3(l - 1, 0, positions)) {
+            v = GreasedLineMesh._CopyV3(1, positions);
+        } else {
+            v = GreasedLineMesh._CopyV3(l - 1, positions);
+        }
+        next.push(v[0], v[1], v[2]);
+        next.push(v[0], v[1], v[2]);
+
+        return {
+            previous,
+            next,
+            uvs,
+            side,
+        };
+    }
+
+    private _drawLine() {

```

Comment:
Not sure `_drawLine` is the best name, I would go for something like `_createVertexBuffers` instead.